pinax/ratings/tests/test_views.py

```diff
@@ -0,0 +1,98 @@
+from django.contrib.contenttypes.models import ContentType
+
+from .models import Car
+from .test import TestCase
+
+
+class RateViewTest(TestCase):
+    # tests for the RateView
+
+    def setUp(self):
+        self.test_user = self.make_user(username="test_user")
+        self.forester = Car.objects.create(name="Subaru Forester")
+
+    def test_post_rating_with_invalid_content_type_id(self):
+        # test line with code snippet:
+        #   get_object_or_404(ContentType, pk=self.kwargs.get("content_type_id"))
+        #
+        with self.login(self.test_user):
+            response = self.post(
+                "pinax_ratings:rate",
+                content_type_id=400,
+                object_id=1,
+                data={
+                    "rating": 3,
+                    "category": "handling"
+                },
+            )
+            self.response_404(response)
+
+    def test_post_rating_with_invalid_object_id(self):
+        # test line with code snippet:
+        #   get_object_or_404(ct.model_class(), pk=self.kwargs.get("object_id"))
+        #
+        with self.login(self.test_user):
+            response = self.post(
+                "pinax_ratings:rate",
+                content_type_id=ContentType.objects.get_for_model(self.forester).pk,
+                object_id=100,
+                data={
+                    "rating": 3,
+                    "category": "handling"
+                },
+            )
+            self.response_404(response)
+
+    def test_post_rating_with_invalid_category(self):
+        # tests the condition:
+        #   if category and cat_choice is None:
+        #
+        with self.login(self.test_user):
+            response = self.post(
+                "pinax_ratings:rate",
+                content_type_id=ContentType.objects.get_for_model(self.forester).pk,
+                object_id=self.forester.pk,
+                data={
+                    "rating": 4,
+                    "category": "non-existing-category"
+                },
+            )
+            self.response_403(response)
+
+    def test_post_rating_with_invalid_rating(self):
+        # tests the condition:
+        #   if rating_input not in range(NUM_OF_RATINGS + 1):
+        #
+        with self.login(self.test_user):
+            response = self.post(
+                "pinax_ratings:rate",
+                content_type_id=ContentType.objects.get_for_model(self.forester).pk,
+                object_id=self.forester.pk,
+                data={
+                    "rating": 10,
```

Comment:
This works for sure, but a more robust test would obtain `NUM_OF_RATINGS` from test settings and test the two so-called "edge" cases. Verify that a rating value greater than the range allows, and one less than the range allows, are both rejected with a 403.

I will point out that `NUM_OF_RATINGS` is a very poor name and actually is not accurate. To me that name indicates how many different ratings are allowed. If `NUM_OF_RATINGS` is five (5), and ratings start with zero (0), I would expect a valid rating range from 0 to 4, a total of five ratings values. However the code doesn't work that way as seen in https://github.com/pinax/pinax-ratings/blob/master/pinax/ratings/views.py#L33.

Go ahead and adjust the test to prove that a rating of `NUM_OF_RATINGS` + 1 is not accepted, and we'll consider fixing the name/code inconsistency later.

Obtain the value from test settings by following the patterns seen in https://github.com/pinax/pinax-announcements/blob/master/pinax/announcements/tests/tests.py. Find references to `settings` for an example.

Reply:
@grahamu I have added the two edge cases test to verify that a rating value greater than the range allows, and one less than the range allows, are both rejected with a 403. The test now obtains the NUM_OF_RATINGS from the settings